### ISSUE TITLE
sparse_cataegorical_accuracy buggued in Keras LSTM when return_sequences=True

### ISSUE THREAD

**Issue by durandg12** [tensorflow/tensorflow#26699]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): not for the original issue, then I tried some workarounds
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS 10.13.6
- TensorFlow installed from (source or binary): pip install tensorflow
- TensorFlow version (use command below): 1.12.0, 1.13.1 and 2.0.0-dev20190313
- Python version: 3.6.7
- We are using CPU here.

**Describe the current behavior**

When training a simple neural network (one embedding layer, one LSTM layer, one dense layer) with `return_sequences=True`, the computation of sparse_categorical_accuracy crashes. The reason of the crash seems different in tensorflow 1.12.0 from one hand, and in tf 1.13.1 and tf 2.0.0 from another hand.

The setting is simple and common for each tensorflow version. See the notebooks in [this dedicated repository](https://github.com/durandg12/keras-lstm-issue) for the complete code. I generate 640 sequences of 5 random integers between 0 and 11. And I want to train an LSTM to predict the 5th element of a sequence given the 4 previous ones, with a batch size of 64. Hence the code looks like this:

```
sequence_length = 4
number_of_categories = 12
BATCH_SIZE = 64
total_examples = 640
steps_per_epoch = total_examples // BATCH_SIZE
raw_data = np.random.randint(0, number_of_categories, size=(total_examples, sequence_length+1))
raw_dataset = tf.data.Dataset.from_tensor_slices(raw_data)
```

**Describe the expected behavior**

The training should not crash because of the accuracy computation. No matter the tensorflow version, note that the training does not crash and the accuracy is properly computed when `return_sequences=False`.

**Code to reproduce the issue**

Please find all the code I used in the notebooks that I have put in [this dedicated repository](https://github.com/durandg12/keras-lstm-issue).

**Other info / logs**

For tensorflow 1.12.0, the error I get is:`Incompatible shapes: [64] vs. [64,4] [Op:Equal]`. `y_true` and `y_pred` should both be of shape `[64,4]` so I quickly figured out where the problem comes from. In the code of `sparse_categorical_accuracy` [found here](https://github.com/tensorflow/tensorflow/blob/r1.12/tensorflow/python/keras/metrics.py), the first line `y_true = math_ops.reduce_max(y_true, axis=-1)` shrinks one dimension of `y_true` for apparently no reason. Hence, with a custom function that simply removes this line, training succeeds, with both `return_sequences=True` and `return_sequences=False`. See [the attached notebook](https://github.com/durandg12/keras-lstm-issue/blob/master/Fail%20LSTM%201.12.0.ipynb).

For tensorflow 1.13.1 and 2.0.0 the situation is more unclear and I couldn't find a workaround this time. This time, the error message is:

```
Can not squeeze dim[1], expected a dimension of 1, got 4
	 [[{{node metrics/sparse_categorical_accuracy_tf13/Squeeze}}]] [Op:StatefulPartitionedCall]
```

and this seems to be due to the lines

```
if (len(K.int_shape(y_true)) == len(K.int_shape(y_pred))):
    y_true = array_ops.squeeze(y_true, [-1])
```
in the code of `sparse_categorical_accuracy` [found here](https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/python/keras/metrics.py) [and here](https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/keras/metrics.py). `y_true` seems to have the expected `[64,4]` shape so we cannot squeeze it on its dimension `1`. But I don't get why we would want to squeeze it in the first place. Looking at the comment line `# If the shape of y_true is (num_samples, 1), squeeze to (num_samples,)` in the code, it is clear that something is not working as intended. My attempt to workaround was to comment the previous lines, so as to use the same workaround as I did with tensorflow 1.12.0. But this time it didn't work for an obscure reason that I can't really understand. The error message I get is:

```
Invalid reduction dimension (2 for input with 2 dimension(s)
	 [[{{node metrics_1/sparse_categorical_accuracy_tf13_v2/Sum}}]] [Op:StatefulPartitionedCall]
```
There seems to be an error with the dimensions of the input data of `sparse_categorical_accuracy`. But `y_true` should be of shape `[64,4]`, and `y_pred`should be of shape `[64,4,12]` before applying ` y_pred = math_ops.argmax(y_pred, axis=-1)` and of shape `[64,4]` after.

So I made an experiment where I replaced the sparse categorical accuracy function by a function simply computing the rank of `y_true` (this is the function `ndim_y_true` in the attached notebooks) and I found even more confusing results. Whereas for 1.12.0, `ndim_y_true` yields `2` as expected for a `[64,4]` tensor, in 1.13.1 and 2.0.0 it yields `3`. This does not make sense to me, especially since when trying to apply `y_true = array_ops.squeeze(y_true, [-1])`, the error message properly aknowledge that the last dimension is dimension `1` (recall the `Can not squeeze dim[1]` above).

When `return_sequences=False`, `ndim_y_true` also yields the strange result of `2` (this time, in the three versions of tensorflow I tried), while I expect `y_true` to be of shape `[64,]` in this case.

As a side note, it seems that adding a `TimeDistributed` layer to the `Dense` layer or not when `return_sequences=True` does not change anything.

**Comment by mogeid** [comment]
I think the problem is somewhere else. If you import keras.metrics.sparse_caategorical_accuracy, you get the same error. Oddly enough, if you literally copy the code for keras.metrics.sparse_categorical_accuracy and put it in your own project, it will run fine. So it seems to me the problem happens when the model is compiled.

**Comment by durandg12** [comment]
> I think the problem is somewhere else. If you import keras.metrics.sparse_caategorical_accuracy, you get the same error. Oddly enough, if you literally copy the code for keras.metrics.sparse_categorical_accuracy and put it in your own project, it will run fine. So it seems to me the problem happens when the model is compiled.

In tf 1.13.1 there is indeed something odd happening at model compilation, because of the value `3` returned by `ndim_y_true`.

Please detail your answer with the version of tf and of keras you used.

**Comment by durandg12** [comment]
I solved the issue (in all tensorflow verisons) by adding a dimension at the end of my target data. In the code of the notebooks I changed the last line of

```
def split_input_target(chunk, return_sequences):
    input_data = chunk[:-1]
    if return_sequences:
        #the target is the whole sequence, shifted by 1
        target_data = chunk[1:]
    else:
        #the target is only the last element
        target_data = chunk[-1]
    return input_data, target_data
```
in `return input_data, expand_dims(target_data, -1)`.

So in sparse encoding, Keras expects the target data to have 3 dimensions, as with one-hot encoding.
